avocado/utils/dmesg.py

```diff
@@ -115,3 +122,19 @@ def collect_errors_by_level(output_file=None, level_check=5):
             err += " Please check  dmesg log in debug log."
             LOGGER.debug(dmsg_log)
         raise DmesgError("Test is failed {}".format(err))
+
+
+def skip_dmesg_error(dmesg_errors, skip_dmesg):
+    """Remove dmesg error messages which want to skip.
+
+    :dmesg_errors: dmesg error messages from which want to filter
+    :type dmesg_errors: string
+    :skip_dmesg: list of dmesg error messages for remove
+    :type skip_dmesg: string
+    """
+    for dmesg in skip_dmesg.split(','):
+        for line in dmesg_errors.splitlines():
+            if dmesg in line:
+                dmesg_errors = dmesg_errors.replace(line, "")
+    dmesg_errors = re.sub(r'\s', '', dmesg_errors)
+    return dmesg_errors
```

Comment:
After reading your comments and the new docstrings again, I think this method needs some refactoring:

1. IIUC, the order of for loops is not efficient, considering that `skip_dmesg` are going to be a smaller list than the number of lines in a dmesg output. 
2. You are using `replace` in order to remove an entire line. This might have some false positives depending on the strings there.
3. You are removing all new lines (`\s`) making the buffer a linear buffer. This is more than what this method is describing. Also your caller method (`collect_errors_by_level()`) will have different output results, depending on the `skip_error` parameter.
4. This code is not only focused on "errors". this could be generic, because is basically doing some filters. So could be useful for other callers. I would recommend removing the "error" string here. 
5. We are inside a python code, would be nice to use proper types (like lists), instead of a "fake string list"
6. When dealing with a buffer like stdouts its is better to handle as bytes instead of strings. Also `process.run()` will return bytes instead of string.

So, again, IIUC, you are doing a `dmesg | egrep -v "some|words|here"`.  If my understanding is correct, IMO this could be an alternative method:

```python
def skip_dmesg_error(dmesg_errors, skip_dmesg):
    """Remove dmesg error messages which want to skip.

    :dmesg_errors: dmesg error messages from which want to filter
    :type dmesg_errors: string
    :skip_dmesg: list of dmesg error messages for remove
    :type skip_dmesg: string
    """
    for dmesg in skip_dmesg.split(','):
        for line in dmesg_errors.splitlines():
            if dmesg in line:
                dmesg_errors = dmesg_errors.replace(line, "")
    dmesg_errors = re.sub(r'\s', '', dmesg_errors)
    return dmesg_errors
```

Of course, we need to change the caller, too:

```diff
-def collect_errors_by_level(output_file=None, level_check=5):
+def collect_errors_by_level(output_file=None, level_check=5, skip_error=None):
     """Verify dmesg having severity level of OS issue(s).
 
     :param output_file: The file used to save dmesg
@@ -97,6 +98,8 @@ def collect_errors_by_level(output_file=None, level_check=5):
                         4 - emerg,alert,crit,err
                         5 - emerg,alert,crit,err,warn
     :type level_check: int
+    :skip_error: dmesg error messages which want skip
+    :type skip_error: string
     """
     if not isinstance(level_check, int):
         raise DmesgError("level_check param should be integer")
@@ -106,7 +109,11 @@ def collect_errors_by_level(output_file=None, level_check=5):
                       verbose=False, shell=True)
     if out.exit_status == 0:
         err = "Found failures in dmesg"
-        dmsg_log = "dmesg log:\n%s" % out.stdout_text
+        if skip_error:
+            dmsg_log = skip_dmesg_error(out.stdout_text, skip_error)
+        else:
+            dmsg_log = "dmesg log:\n%s" % out.stdout_text
+    if dmsg_log:
         if output_file:
             with open(output_file, "w+") as log_f:
                 log_f.write(dmsg_log)
@@ -115,3 +122,19 @@ def collect_errors_by_level(output_file=None, level_check=5):
             err += " Please check  dmesg log in debug log."
             LOGGER.debug(dmsg_log)
         raise DmesgError("Test is failed {}".format(err))
```

What do you think?

Reply:
"replace" i am using to remove entire line,in which line contain the skip dmesg. in whole line with messages there is date time etc. things also there.so i am using replace for that. and after remove the message lines there will create empty line so using "\s" for that.